### PR TITLE
fix(apigatewayv2): Invalid mapping key value

### PR DESCRIPTION
**Pull request by pahud**
- `mappingKey` is now optional
-  undefinied `mappingKey` indicates the root base path mapping for custom domain
-  ensure the resource dependency for the api mapping and stage

Fixes: https://github.com/aws/aws-cdk/issues/8983

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api-mapping.ts [hunk 1]
```diff
@@ -61,18 +61,44 @@ export class HttpApiMapping extends Resource implements IApiMapping {
...
     super(scope, id);
 
+    if ((!props.stage?.stageName) && !props.api.defaultStage) {
+      throw new Error('stage is required if default stage is not available');
+    }
+
+    const paramRe = '^[a-zA-Z0-9]+[$-_.+!*\'\(\),]?$';
+    if (props.apiMappingKey && !new RegExp(paramRe).test(props.apiMappingKey)) {
+      throw new Error('An ApiMapping key may contain only letters, numbers and one of $-_.+!*\'(),');
+    }
```
Comment (ericzbeard) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/api-mapping.ts:79]: Can you add a unit test to make sure this regex behaves exactly as you expect?

Comment (pahud) [reply packages/@aws-cdk/aws-apigatewayv2/lib/http/api-mapping.ts:79]: Thanks for the review.

I just added tests on `ApiMappingKey` against the following conditions:

'/'
'/foo'
'foo/bar'
'foo/'
'^foo'
'foo.*$'

**Comment by pahud** [comment]
I need update my regex pattern. WIP now.

**Comment by pahud** [comment]
Hi @ericzbeard 

I am ready now. Please take a look again.

**Pull request merged**